### PR TITLE
base64

### PR DESCRIPTION
**Pull request by neotheprogramist**

### CODE REVIEW

**Review by tarrencev — APPROVED** [review]
looks great

**Comment by LucasLvy** [comment]
How about a generic encoder trait and then you implement it for base64 ?

**Comment by neotheprogramist** [comment]
@LucasLvy Could you take a look at the code now?
If you have any further suggestions lmk :)

**Comment by gaetbout** [comment]
Could you update the main Readme ## Features section?

File: alexandria/encoding/src/base64.cairo [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+            if *data[data.len() - 2] == '=' {
+                p += 1;
+            }
+            decode_loop(p, ref data, 0, ref result);
+        }
+        result
+    }
+}
+
+impl u8Base64UrlDecoder of Decoder<u8> {
```
Comment (gaetbout) [inline alexandria/encoding/src/base64.cairo:73]: As both decoder are the same, I wouldn't do two just for the sake of having one that suit each name.
But if you think it makes sense I guess we can keep it, no strong opinion

File: alexandria/encoding/src/math.cairo [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    }
+}
+
+fn shl(x: u128, n: u128) -> u128 {
+    x * fpow(2, n)
+}
+
+fn shr(x: u128, n: u128) -> u128 {
+    x / fpow(2, n)
+}
```
[the remark this answers is not in the document]
Comment (neotheprogramist) [reply alexandria/encoding/src/math.cairo:20]: This is the import problem I was writing about in the group.
I don't know how to import functions between crates correctly, so I left that doubled but working

[the remark this answers is not in the document]
Comment (gaetbout) [reply alexandria/encoding/src/math.cairo:20]: In the encoding scarb.toml put
```toml
[dependencies]
quaireaux_math = { path = "../math/" }
```
in the encoding cairo_project.toml put:
```
quaireaux_math = "../math/src"
```
Move all to the math file and update local import of sha256
then in encoding crate you can use math fn like:
```
use quaireaux_math::{count_digits_of_base, pow};
```

**Comment by gaetbout** [comment]
calling @LucasLvy for final review

File: alexandria/encoding/cairo_project.toml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+[crate_roots]
+alexandria_encoding = "src"
+alexandria_math = "../math/src"
```
Comment (LucasLvy) [inline alexandria/encoding/cairo_project.toml:3]: Does it need to be here ?

File: alexandria/encoding/src/base64.cairo [hunk 4]
```diff
@@ -0,0 +1,246 @@
+use core::clone::Clone;
+use array::ArrayTrait;
+use integer::{upcast, downcast};
+use option::OptionTrait;
+use debug::PrintTrait;
+use alexandria_math::math::{shl, shr, U6_MAX, U8_MAX};
+
+trait Encoder<T> {
+    fn encode(data: Array<T>) -> Array<u8>;
```
Comment (LucasLvy) [inline alexandria/encoding/src/base64.cairo:9]:
```suggestion
    fn encode(data: T) -> Array<u8>;
```
How about this ? so you can encode anything ?

File: alexandria/encoding/src/base64.cairo [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+}
+
+trait Decoder<T> {
+    fn decode(data: Array<T>) -> Array<u8>;
```
Comment (LucasLvy) [inline alexandria/encoding/src/base64.cairo:13]:
```suggestion
    fn decode(data: T) -> Array<u8>;
```
How about this ? so you can decode anything ?

File: alexandria/encoding/src/base64.cairo [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        }
+        if *data[data.len() - 2] == '=' {
+            p += 1;
+        }
+        decode_loop(p, ref data, 0, ref result);
+    }
+    result
+}
+
+fn decode_loop(p: u8, ref data: Array<u8>, d: usize, ref result: Array<u8>) {
```
Comment (LucasLvy) [inline alexandria/encoding/src/base64.cairo:81]: you can use a loop there (maybe you prefer recursive ?)

File: alexandria/encoding/src/base64.cairo [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+    } else if (x == '_') {
+        63
+    } else if (x <= 'z') {
+        (x - 'a') + 26
+    } else {
+        0
+    }
+}
+
+fn encode_loop(
```
Comment (LucasLvy) [inline alexandria/encoding/src/base64.cairo:141]: you can use a loop there (maybe you prefer recursive ?)

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: alexandria/encoding/cairo_project.toml (hunk 3 above)
Comment (gaetbout) [reply alexandria/encoding/cairo_project.toml:3]: It is useful as it uses some code from math such as pow and shl and shr

File: alexandria/encoding/src/base64.cairo (hunk 6 above)
Comment (neotheprogramist) [reply alexandria/encoding/src/base64.cairo:81]: With the current state of Cairo, I prefer recursion

**Pull request merged**